payment-processor/src/main/kotlin/org/ostelco/prime/paymentprocessor/StripePaymentProcessor.kt

```diff
@@ -223,63 +265,75 @@ class StripePaymentProcessor : PaymentProcessor {
             either("Failed to subscribe customer $stripeCustomerId to plan $planId") {
                 val item = mapOf("plan" to planId)
                 val taxRates = getTaxRatesForTaxRegionId(taxRegionId)
                         .fold(
                                 { emptyList<TaxRateInfo>() },
                                 { it }
                         )
-                val subscriptionParams = mapOf(
+                val trialEndInEpochSeconds = ofEpochMilliToSecond(trialEnd)
+                val params = mapOf(
                         "customer" to stripeCustomerId,
                         "items" to mapOf("0" to item),
-                        *(if (trialEnd > Instant.now().epochSecond)
-                            arrayOf("trial_end" to trialEnd.toString())
+                        *(if (trialEndInEpochSeconds > Instant.now().epochSecond)
+                            arrayOf("trial_end" to trialEndInEpochSeconds.toString())
                         else
                             arrayOf("expand" to arrayOf("latest_invoice.payment_intent"))),
                         *(if (taxRates.isNotEmpty())
                             arrayOf("default_tax_rates" to taxRates.map { it.id })
-                        else arrayOf()))
-                val subscription = Subscription.create(subscriptionParams)
-                val status = subscriptionStatus(subscription)
-                SubscriptionDetailsInfo(id = subscription.id,
-                        status = status.first,
-                        invoiceId = status.second,
-                        chargeId = status.third,
-                        created = Instant.ofEpochSecond(subscription.created).toEpochMilli(),
-                        trialEnd = subscription.trialEnd ?: 0L)
+                        else arrayOf()),
+                        "payment_behavior" to "allow_incomplete",
+                        "prorate" to true)
+                getSubscriptionPaymentInfo(Subscription.create(params))
             }
 
-    private fun subscriptionStatus(subscription: Subscription): Triple<PaymentStatus, String, String> {
+    private fun getSubscriptionPaymentInfo(subscription: Subscription): SubscriptionPaymentInfo {
         val invoice = subscription.latestInvoiceObject
         val intent = invoice?.paymentIntentObject
 
-        return when (subscription.status) {
-            "active", "incomplete" -> {
-                if (intent != null)
-                    Triple(when (intent.status) {
+        return if (subscription.status == "trialing") {
+            PaymentStatus.TRIAL_START
+        } else {
+            require(intent != null) {
+                "Expected an invoice intent with invoice ${invoice.id} for subscription ${subscription.id} " +
+                        "with status ${subscription.status}"
+            }
+            when (subscription.status) {
+                "active" -> {
+                    when (intent.status) {
```

Comment:
Please avoid using magic strings like "succeed", "incomplete" etc. If these refer to protocol options, then please use enumerations with strings in them to identify the different protocol options.   Naked (and thus magic) strings obscure what's going on, and are also difficult to keep track of if used more than one place in the codebase (and it's hard to know if it's used more than once since text search is necessary, it's much easier to  jump to the enumeration and then see where it's used).